services/worker/src/worker/job_runners/split/descriptive_stats.py

```diff
@@ -0,0 +1,288 @@
+from worker.utils import CompleteJobResult
+
+DECIMALS = 5
+
+INTEGER_DTYPES = ["int8", "int16", "int32", "int64"]
+FLOAT_DTYPES = ["float16", "float32", "float64"]
+
+
+SplitDescriptiveStatsJobRunnerErrorCode = Literal["PreviousStepFormatError"]
+
+
+class Histogram(TypedDict):
+    hist: List[int]
+    bin_edges: List[float]
+
+
+class NumericalStatsItem(TypedDict):
+    nan_count: int
+    nan_prop: float
+    min: float
+    max: float
+    mean: float
+    median: float
+    std: float
+    histogram: Histogram
+
+
+class CategoricalStatsItem(TypedDict):
+    nan_count: int
+    nan_prop: float
+    n_unique: int
+    frequencies: Dict[str, int]
+
+
+class StatsPerColumnItem(TypedDict):
+    column_name: str
+    column_type: str
+    column_dtype: Optional[str]
+    column_stats: Union[NumericalStatsItem, CategoricalStatsItem]
+
+
+class SplitDescriptiveStatsResponse(TypedDict):
+    num_examples: int
+    stats: List[StatsPerColumnItem]
+
+
+def compute_histogram(
+    con: duckdb.DuckDBPyConnection,
+    column_name: str,
+    urls: List[str],
+    bin_size: int,
+) -> Histogram:
+    hist_query = f"""
+    SELECT FLOOR("{column_name}"/{bin_size})*{bin_size}, COUNT(*) FILTER (WHERE {column_name} IS NOT NULL)
+     FROM read_parquet({urls}) GROUP BY 1 ORDER BY 1;
+    """
+    logging.debug(f"Compute histogram for {column_name}")
+    bins, hist = zip(*con.sql(hist_query).fetchall())  # 2 tuples
+    bins, hist = list(bins), list(hist)
+    if None in bins:  # if there are None values in column
+        # it should be always the last element since we order in query but just in case
+        none_idx = bins.index(None)
+        bins.pop(none_idx)
+        hist.pop(none_idx)
+    bins = np.round(bins, DECIMALS).tolist()
+    return Histogram(hist=hist, bin_edges=bins)
+
+
+def compute_numerical_stats(
+    con: duckdb.DuckDBPyConnection,
+    column_name: str,
+    urls: List[str],
+    n_bins: int,
+    n_samples: int,
+    dtype: str,
+) -> NumericalStatsItem:
+    query = f"""
+    SELECT min({column_name}), max({column_name}), mean({column_name}), median({column_name}),
+     stddev_samp({column_name}) FROM read_parquet({urls});
+    """
+    minimum, maximum, mean, median, std = duckdb.query(query).fetchall()[0]
+    if dtype in FLOAT_DTYPES:
+        bin_size = np.round((maximum - minimum) / n_bins, decimals=DECIMALS).item()
+        minimum, maximum, mean, median, std = np.round([minimum, maximum, mean, median, std], DECIMALS).tolist()
+    elif dtype in INTEGER_DTYPES:
+        if maximum - minimum < n_bins:
+            bin_size = 1
+        else:
+            bin_size = int(np.round((maximum - minimum) / n_bins))
+        mean, median, std = np.round([mean, median, std], DECIMALS).tolist()
+    else:
+        raise ValueError("Incorrect dtype, only integers and float are allowed. ")
+    nan_query = f"SELECT COUNT(*) FILTER (WHERE {column_name} IS NULL) FROM read_parquet({urls});"
```

Comment:
Not sure if this will apply to all datasets but I did a small test of different performances for:
`EXPLAIN ANALYZE SELECT COUNT(*) FROM read_parquet("https://huggingface.co/datasets/glue/resolve/refs%2Fconvert%2Fparquet/ax/glue-test.parquet") WHERE premise IS NULL
`and
`EXPLAIN ANALYZE SELECT COUNT(*) FROM read_parquet("https://huggingface.co/datasets/glue/resolve/refs%2Fconvert%2Fparquet/ax/glue-test.parquet") WHERE premise IS NULL;
`
 And looks like the second one is faster:
First result: Total Time: 0.382s
Second result: Total Time: 0.0001s

Maybe EXPLAIN ANALYZE command can help you identify critical steps https://duckdb.org/dev/profiling.html